File: 1352-product-of-the-last-k-numbers/1352-product-of-the-last-k-numbers.go
type ProductOfNumbers struct {
    list []int
    size int
}


func Constructor() ProductOfNumbers {
    return ProductOfNumbers {
        list : []int{},
        size : 0,
    }
}


func (this *ProductOfNumbers) Add(num int)  {
    if num == 0 {
        this.list = this.list[:0]
        return
    }
    if len(this.list) == 0 {
        this.list = append(this.list, num)
        return
    }
    last := this.list[len(this.list) - 1]
    this.list = append(this.list, last * num)
}


func (this *ProductOfNumbers) GetProduct(k int) int {
    length := len(this.list)
    last := length - 1
    if k > length {
        if length == this.size {
            return this.list[last]
        } else {
            return 0
        }
    }
    if k == length {
        return this.list[last]
    }
    return this.list[last] / this.list[last - k]
}


/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(num);
 * param_2 := obj.GetProduct(k);
 */

File: 1352-product-of-the-last-k-numbers/1352-product-of-the-last-k-numbers.ts
class ProductOfNumbers {
    private list: number[];
    private size: number;
    constructor() {
        this.list = [];
        this.size = 0;
    }

    add(num: number): void {
        this.size++;
        switch (true) {
            case num === 0:
                this.list.length = 0;
                break;
            case this.list.length === 0:
                this.list.push(num);
                break;
            default:
                const last = this.list[this.list.length - 1];
                this.list.push(last * num);
                break;
        }
        return null;
    }

    getProduct(k: number): number {
        const length = this.list.length;
        const last = length - 1;
        switch (true) {
            case k > length:
                return length === this.size ? this.list[last] : 0
            case k === length:
                return this.list[last]
            default:
                return this.list[last] / this.list[last - k]
        }
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * var obj = new ProductOfNumbers()
 * obj.add(num)
 * var param_2 = obj.getProduct(k)
 */